### PR TITLE
perf(moe): add scoped partial CUDA graphs

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do?

Adds Transformer Engine scoped CUDA Graph capture for fixed-shape attention and MoE preprocessing, with one focused extension in this revision: `attn` now means the whole attention module, matching Megatron Core naming.

Capture scopes:

- `attn`: whole attention (`torch` Q/K/V/O projections + `torch` RMSNorm + unfused RoPE + TE DPA)
- `te_dpa`: the previous narrow, parameterless TE fused-attention boundary
- `moe_router`: learned router selection/probability core; gate projection remains eager
- `moe_preprocess`: fused HybridEP top-k-to-multihot preprocessing

Dynamic dispatch, expert compute, and combine remain eager. Capture starts after one complete eager training step. Replay is guarded by pytree, tensor metadata, aliasing, non-tensor controls, parameter metadata/local storage, and captured buffer storage; mismatches fall back to eager execution.

Capture is transactional: adapters are staged first, every staged graph is reset if any later target or input variant fails, and forwards are patched only after the complete capture succeeds. Optional TE loading uses the repository safe-import path and non-empty zero-pointer parameter/buffer storage is rejected before capture.

The benchmark recipe now uses the same capture/cleanup lifecycle as the training recipe and reports capture/replay/fallback counts. It also removes an unclosed `emit_nvtx()` context that installed persistent RecordFunction callbacks and broke DTensor/FSDP profiling.

## Configuration

```yaml
model:
  backend:
    attn: te
    linear: torch
    rms_norm: torch
    rope_fusion: false
    cuda_graph_modules: [attn]
```

The list follows Megatron Core's module naming. `attn` and `te_dpa` are mutually exclusive. Existing experimental users who want the old narrow behavior should rename `attn` to `te_dpa`.

Whole-attention capture currently requires BF16 DPA, PyTorch linear/RMSNorm, unfused RoPE, fixed input contracts, and no activation checkpointing or pipeline parallelism.

## Why parameters are explicit graph inputs

Whole attention owns PyTorch projection and norm parameters. Allowing TE to discover those as static module state failed single-GPU backward capture because an `AccumulateGrad` node on the legacy stream introduced an invalid capture dependency. Passing parameters explicitly avoids that failure and also lets FSDP2 replay use the currently materialized parameter storage.

For FSDP2, the owning block is unsharded before capture and always resharded in `finally`. Replay validates that the FSDP forward hook materialized stable local parameters; nested/unmaterialized ownership fails closed.

## Validation

Local:

- affected partial-graph/training-loop unit tests: `218 passed, 5 skipped`
- Ruff format/check and `git diff --check`: passed

H100, PyTorch `2.12.0a0+0291f960b6.nv26.04.48445190`, Transformer Engine `2.16.0+4220403e`:

- single GPU, mixed PyTorch + TE attention: `captured=1`, `replayed=3`, `fallback=0`, maximum absolute eager/graph error `0.0` (job 5573998)
- 2-rank FSDP2 with rank-dependent inputs and post-capture reshard assertion: `captured=1`, `replayed=3`, `fallback=0`, maximum absolute eager/graph error `0.0` (job 5573999)

The runner compares outputs, loss, input gradients, parameter gradients, and optimizer-updated parameters. It is a standalone H100 functional runner, not a CI-registered pytest.

## Real-model E2E performance

The representative run uses the real `Qwen/Qwen3-30B-A3B` checkpoint (30.53B trainable parameters) on 8 H100s with BF16 FSDP2 + EP8 + DeepEP, sequence length 1,024, local/global batch 1/8, no activation checkpointing, and a mock token dataset. This is a complete forward, backward, and Adam optimizer workload. All 48 whole-attention modules are captured; projections, RMSNorm, and RoPE are PyTorch operations and DPA is TE.

The long comparison used fresh processes in symmetric G-E-E-G order. Each process ran 10 warmup and 100 measured steps without a profiler:

| Position | Mode | Mean step |
|---:|---|---:|
| 1 | graph | 424.871 ms |
| 2 | eager | 412.406 ms |
| 3 | eager | 406.666 ms |
| 4 | graph | 375.145 ms |

Across both repeats:

- mean latency: 400.008 ms graph vs 409.536 ms eager, **2.33% lower** (2.38% throughput increase)
- median latency: 340.647 ms graph vs 360.225 ms eager, **5.43% lower**
- 10% trimmed mean: 346.376 ms graph vs 361.522 ms eager, **4.19% lower**
- graph lifecycle per graph run: `captured=48`, `replayed=5232`, `fallback=0`
- peak memory: 38.09 GiB graph vs 31.79 GiB eager, **+6.30 GiB (+19.8%)**
- final losses remained aligned (`6.9523`/`6.9522` and `6.9527`/`6.9529` graph/eager); the exact parity harness above is the stronger correctness check

The primary result is the all-step mean: this is a modest, positive speedup with substantial memory cost. Median and trimmed means show that the normal steady-state step improves more than the mean, while uncaptured DeepEP/NCCL tail latency reduces the aggregate gain.

### Nsight diagnosis

Three steady-state steps were profiled across 8 ranks. A second graph profile used `--cuda-graph-trace=node` so graph child kernels were not hidden by graph-level tracing. Profiler wall time is intentionally not used as a throughput result.

| Trace metric (8 ranks × 3 steps) | Eager | Graph | Interpretation |
|---|---:|---:|---|
| GPU kernel executions | 264,903 | 264,697 | essentially unchanged (-0.08%); graph does not skip compute |
| CUDA runtime API calls | 742,518 | 584,392 | 21.3% fewer host/runtime calls |
| `cudaLaunchKernel` calls | 206,552 | 162,280 | 21.4% fewer ordinary launches |
| `cudaGraphLaunch` calls | 0 | 2,304 | 48 forward + 48 backward replays per rank/step |
| D2D copy volume | 89.57 MiB | 50,784.38 MiB | approximately +2.11 GiB/rank/step for static inputs, primarily explicit attention parameters |

DeepEP dispatch/combine and NCCL kernel counts are unchanged. The graph therefore saves host launch overhead inside attention, but it neither captures the dynamic MoE communication path nor removes its tail latency. Explicit parameter/static-input copies and persistent buffers explain the added bandwidth and memory cost.

## Other real-model findings

- `openai/gpt-oss-20b`, 8K, 8 H100s needs full activation checkpointing to fit this setup. The retained narrow `te_dpa` scope reached `captured=2`, `replayed=50`, `fallback=0`; a 20-step E-G-E run measured 2.128/2.068/2.043 s, which is parity within the 4.1% eager-to-eager variation, with 45.99/40.70 GiB graph/eager memory. The new whole-`attn` scope intentionally rejects activation checkpointing, while the no-AC 8K configuration OOMs, so GPT-OSS 20B/8K is not yet a supported whole-attention performance target.
- The real `Qwen/Qwen2.5-0.5B` dense model at 8K currently fails whole-attention capture during TE helper backward warmup with saved-tensor reuse. Removing TE warmup instead fails CUDA stream capture for `AccumulateGrad`. This is a real architecture gap, not a benchmark result, and requires a TE helper change or a purpose-built autograd boundary.

## Performance conclusion and remaining scope

Whole-attention capture is now demonstrated end-to-end under FSDP2 with a 2.33% mean latency reduction on Qwen3-30B-A3B. The +19.8% peak-memory increase is too large to enable it by default. The highest-value follow-up is to avoid copying/replicating the whole attention parameter surface, either through stable FSDP parameter storage registration or a TE-native local/TransformerLayer boundary. Capturing dynamic DeepEP dispatch/combine remains unsafe without a fixed communication contract.

Still out of scope: full-iteration graphs, expert graphs/token bucketing, FP8 DPA, activation checkpointing for whole attention, PP/VPP, dynamic sequence shapes, and a production performance recipe.

## Attribution

The original scoped partial-graph implementation is based on Hemil Desai's work, especially commit `f3a6bed2`; Hemil remains credited as co-author on the base commit.

Main tracking issue: #1027.

### CODE REVIEW

File: nemo_automodel/components/cuda_graphs/partial.py [hunk 1]
```diff
@@ -0,0 +1,1374 @@
...
+import torch
+from torch import nn
+from torch.utils._pytree import tree_flatten, tree_unflatten
+
+logger = logging.getLogger(__name__)
+
+_Canonicalizer = Callable[[tuple[Any, ...], dict[str, Any]], tuple[tuple[Any, ...], dict[str, Any]]]
+
+
+def _get_make_graphed_callables() -> Callable[..., Any]:
```
Comment (yaoyu-33) [inline nemo_automodel/components/cuda_graphs/partial.py:41]: feels not necessary to make such function, can just import and use later.

Comment (yaoyu-33) [reply nemo_automodel/components/cuda_graphs/partial.py:41]: Kept this helper intentionally after the follow-up review. It is the narrow lazy-loading boundary for the optional Transformer Engine dependency: it now uses `safe_import_from()`, raises an actionable error when TE does not expose the graph helper, and gives the unit tests a focused seam without importing TE at module load time. A direct import would make this recipe module fail to load in CPU/no-TE environments. Added coverage for the unavailable-helper path.

**Comment by akoumpa** [comment]
I have read the DCO and hereby sign off the commits in this pull request.

**Review by akoumpa — APPROVED** [review]

File: nemo_automodel/components/moe/layers.py [hunk 2]
```diff
@@ -292,38 +305,12 @@ def __init__(
...
-            x_compute = x.to(dtype=self.gate_precision)
-            weight = self.weight.to(dtype=self.gate_precision)
-            bias = self.bias.to(dtype=self.gate_precision) if self.bias is not None else None
-        else:
-            x_compute = x
-            weight = self.weight.to(dtype=x.dtype)
-            bias = self.bias.to(dtype=x.dtype) if self.bias is not None else None
-
-        scores = F.linear(x_compute, weight, bias=bias)
+    def _route_scores(self, scores: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor, Optional[torch.Tensor]]:
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply nemo_automodel/components/moe/layers.py:311]: Addressed in NVIDIA-NeMo/Automodel@2b3b76129af109bd3111a6cdd25189f3bf797c9b on the current stack root (#2932). Both `_GateRoutingCore.forward` and `Gate._route_scores` now document `[tokens, num_experts]` inputs and all `[tokens, top_k]` / `[tokens, num_experts]` outputs.

File: nemo_automodel/components/moe/megatron/token_dispatcher.py [hunk 3]
```diff
@@ -319,6 +320,38 @@ def get_restored_hidden_states_by_experts(self, hidden_states: torch.Tensor) ->
...
 
+class _HybridEPMetadataProcessor(nn.Module):
+    """Fixed-shape conversion from top-k metadata to HybridEP metadata."""
+
+    def __init__(self, *, num_experts: int, permute_fusion: bool):
+        super().__init__()
+        self.num_experts = num_experts
+        self.permute_fusion = permute_fusion
+
+    def forward(self, token_indices: torch.Tensor, token_probs: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply nemo_automodel/components/moe/megatron/token_dispatcher.py:331]: Addressed on the current stack root (#2932). `_HybridEPMetadataProcessor.forward` now has Google-style `Args` and `Returns` covering compact `[tokens, top_k]` metadata and dense `[tokens, num_experts]` outputs, including masked expert ids and gradient-preserving probabilities.
